File: Arrays/majorityElement.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<map>
using namespace std;
int majorityElement_hashing(vector<int>arr){
map<int,int>mpp;
for(int i=0;i<arr.size();i++){
    mpp[arr[i]]++;
}
for(auto it:mpp){
    if(it.second > (arr.size()/2)){
        return it.first;
    }
}
return -1;
}

int majorityElement_moores_voting_algo(vector<int>arr){
    int cnt=0;
    int el;
    for(int i=0;i<arr.size();i++){
        if(cnt==0){
            cnt=1;
            el=arr[i];
        }else if(arr[i]==el){
            cnt++;
        }else{
            cnt--;
        }
    }
    int cnt1=0;
    for(int i=0;i<arr.size();i++){
        if(arr[i]==el) cnt1++;
    }
    if(cnt1>(arr.size()/2)){
        return el;
    }
    return -1;
}

int majorityElement_brute(vector<int>arr){
    for(int i=0;i<arr.size();i++){
            int cnt=0;
        for(int j=0;j<arr.size();j++){
            if(arr[i]==arr[j]){
                cnt++;
            }
        }
        if(cnt>(arr.size()/2)) return arr[i];
    }
    return -1;
}

int main() {
    vector<int> arr = {2, 2, 1, 2, 2, 3, 2};

    cout << "Majority using brute: " << majorityElement_brute(arr) << endl;
    cout << "Majority using hashing: " << majorityElement_hashing(arr) << endl;
    cout << "Majority using Moore's Voting: " << majorityElement_moores_voting_algo(arr) << endl;

    return 0;
}

File: ArrayQ/MajorityElement.java
import java.util.*;

public class MajorityElement {

    // ---------------- Brute Force ----------------
    // Time Complexity: O(N^2)
    // Space Complexity: O(1)
    static int majorityElementBrute(int[] arr) {
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            int cnt = 0;
            for (int j = 0; j < n; j++) {
                if (arr[i] == arr[j]) {
                    cnt++;
                }
            }
            if (cnt > n / 2) return arr[i];
        }
        return -1;
    }

    // ---------------- Hashing ----------------
    // Time Complexity: O(N)
    // Space Complexity: O(N)
    static int majorityElementHashing(int[] arr) {
        HashMap<Integer,Integer>mpp=new HashMap<>();
        for(int i=0;i<arr.length;i++){
            mpp.put(arr[i],mpp.getOrDefault(arr[i],0)+1);
        }
        for(var entry:mpp.entrySet()){
            if(entry.getValue()>arr.length/2){
                return entry.getKey();
            }
        }
        return -1;
    }

    // ---------------- Moore’s Voting Algorithm ----------------
    // Time Complexity: O(N)
    // Space Complexity: O(1)
    static int majorityElementMooresVoting(int[] arr) {
        int cnt = 0;
        int el = 0;

        // Step 1: Find potential candidate
        for (int num : arr) {
            if (cnt == 0) {
                el = num;
                cnt = 1;
            } else if (num == el) {
                cnt++;
            } else {
                cnt--;
            }
        }

        // Step 2: Verify candidate
        int count = 0;
        for (int num : arr) {
            if (num == el) count++;
        }

        if (count > arr.length / 2) return el;
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {2, 2, 1, 2, 2, 3, 2};

        System.out.println("Majority using brute: " + majorityElementBrute(arr));
        System.out.println("Majority using hashing: " + majorityElementHashing(arr));
        System.out.println("Majority using Moore's Voting: " + majorityElementMooresVoting(arr));
    }
}
